### ISSUE TITLE
Issues with GET parameters in OAuth2 redirect URI

### ISSUE THREAD

**Issue by joho1968** [mattermost/mattermost-server#20716]
#### Summary
Issues with GET parameters in OAuth2 redirect URI

#### Steps to reproduce
Supply a redirect_uri containing a GET parameter when MM is called (using MM 7.1.2)

#### Expected behavior
I would expect the callback by MM to be correctly issued

#### Observed behavior (that appears unintentional) 
I see the GET parameters passed by MM to the client in the callback to be "bungled" (see below)

#### My client calling Mattermost

https://mm.server.com/oauth/authorize?state=1234&response_type=code&approval_prompt=auto&redirect_uri=https%3A%2F%2Fmy.test.com%2Fplugins%2Fxplug%2Foauth2.php%3Fxregister%3Dyes&client_id=1234

#### Mattermost issuing the callback
When Mattermost completes the authentication process, the file oauth2.php is called with the following GET parameters:

  'xregister' => 'yes?code=1234',                                               
  'state' => '1234'
  
But it should be

  'xregister' => 'yes',
  'code' => '1234',                                               
  'state' => '1234'
  
No?

#### Other things
To get this to work at all, and not have MM complain about an invalid redirect parameter, I had to add the URI as both https://my.test.com/plugins/xplug/oauth2.php and https://my.test.com/plugins/xplug/oauth2.php?xregister=yes (i.e. two lines).

Am I trying to do something that wasn't intended to work?

(I realize I could have separate redirection handlers on the client side to address this, but that seems "sort of" an inflexible way of doing it.)

**Comment by amyblais** [comment]
@joho1968 Would you be open to raising this issue at https://community.mattermost.com/core/channels/peer-to-peer-help or at https://forum.mattermost.com/? One of our community or staff members will be able to help troubleshoot this.

**Comment by joho1968** [comment]
I've done that now:
https://community.mattermost.com/core/pl/ba6nykiuuidkxpf9ixbrg8z4jr

**Comment by joho1968** [comment]
(By the way, the WordPress plugin now does this as suggested in a reply to my original post:
https://community.mattermost.com/core/pl/ixwmzj1533g9fxqextsrsisx5c)

The plugin is at https://wordpress.org/plugins/cloudbridge-mattermost

**Comment by amyblais** [comment]
Opened: https://mattermost.atlassian.net/browse/MM-46216.

**Comment by plant99** [comment]
Hi @joho1968 , thanks for pointing this out. There could sure be an issue here, which I'll report if I find something.

But the issue isn't parameter mixing up, but why is `xregister` query param showing up in callback when it isn't in registered redirected URI in the first case.

According some references [here](https://www.oauth.com/oauth2-servers/redirect-uris/redirect-uri-registration/) and [here](https://stackoverflow.com/a/55577647), the redirect URL shouldn't change per request except for the `state` parameter. It'd be ideal if you move custom query parameters into state as an encoded string.

Please note again that I'm still looking into it, and will post here if a fix is required.

**Comment by joho1968** [comment]
Well, I guess this could be looked at both ways. But I'm guessing I was doing it wrong. I am now (as noted above) doing it with the state parameter instead and all is well.

**Comment by plant99** [comment]
Hey, apologies if my above message is irrelevant. But it's a valid bug, here's a redirect URL I get when mine has query parameters.

```
 /oauth/callback?something=one&fc=anderchlect?code=ws1yz9c95br7ic6oiftoakasphgrdz7aes6tf9tx6gryb1nuqskc&state=
 ```

**Comment by joho1968** [comment]
No, I don't think your message is irrelevant, it is, IMHO, a bug. What I meant was that if I had followed the "official guide" and used the state parameter, this would not have been an issue. OTOH, then we wouldn't have seen the bug either 😊

**Comment by cpatulea** [comment]
We hit this as well, we built an integration with Mattermost OAuth service provider (https://github.com/FOULAB/matterid) and this stopped working due to the double-escaping (our OAuth app needs the state parameter and the value is now invalid due to the double-escaping).

It was working with version 6.5.2, we upgraded to 8.1.7, and it stopped working. We would like to stay on the 8.1 series because it's an Extended Support Release. Are there any plans to backport the fix to the 8.1 series?
